libkbfs/rekey_fsm.go

```diff
@@ -418,42 +434,30 @@ func NewRekeyFSM(fbo *folderBranchOps) RekeyFSM {
 }
 
 func (m *rekeyFSM) loop() {
-	for e := range m.reqs {
-		if e.eventType == rekeyShutdownEvent {
-			atomic.StoreInt32(&m.reqsClosed, 1)
-			close(m.reqs)
+	for {
+		select {
+		case e := <-m.reqs:
+			next := m.current.reactToEvent(e)
+			m.log.Debug("RekeyFSM transition: %T + %s -> %T",
+				m.current, e, next)
+			m.current = next
+
+			m.triggerCallbacksForTest(e)
+		case <-m.shutdownCh:
+			return
 		}
-
-		next := m.current.reactToEvent(e)
-		m.log.Debug("RekeyFSM transition: %T + %s -> %T",
-			m.current, e, next)
-		m.current = next
-
-		m.triggerCallbacksForTest(e)
 	}
 }
 
 // Event implements RekeyFSM interface for rekeyFSM.
 func (m *rekeyFSM) Event(event RekeyEvent) {
 	select {
 	case m.reqs <- event:
-		return
-	default:
-	}
-	go func() {
-		// We use a spinning loop here to avoid doing a naked `m.reqs <- event`,
-		// which in racy conditions could cause a panic due to sending to closed
-		// channel. The spinning loop is fine here since reactToEvent
-		// implementations of states are supposed to consume events from m.reqs
-		// pretty fast.
-		for atomic.LoadInt32(&m.reqsClosed) == 0 {
-			select {
-			case m.reqs <- event:
-				return
-			default:
-			}
+		if event.eventType == rekeyShutdownEvent {
+			close(m.shutdownCh)
```

Comment:
This was because of the principle that writer is supposed to signal the end. But I just realized this means multiple calls to `close(m.shutdownCh)` could happen multiple times, so will move it into the reader loop.